pytorch_lightning/trainer/properties.py

```diff
@@ -238,16 +239,17 @@ def save_checkpoint(self, filepath, weights_only: bool = False):
     def get_model(self):
         return self.model_connector.get_model()
 
+    @property
+    def lightning_optimizers(self):
+        if self._lightning_optimizers is None:
+            self.convert_to_lightning_optimizers()
+        return self._lightning_optimizers
+
     def __getstate__(self):
-        # unwrap optimizer
-        self.optimizers = [opt._optimizer if is_lightning_optimizer(opt) else opt for opt in self.optimizers]
+        # remove lightning_optimizers
+        self._lightning_optimizers = None
```

Comment:
woun't this affect some other calls?

Reply:
No it shouldn't. 

Here is the flow.

```
    @property
    def lightning_optimizers(self):
        if self._lightning_optimizers is None:
            self.convert_to_lightning_optimizers()
        return self._lightning_optimizers

    def __getstate__(self):
        # remove lightning_optimizers
        self._lightning_optimizers = None
        return self.__dict__
```
We need to `self._lightning_optimizers = None` everytime we apply a deepcopy on Trainer.

When restarting the training.

```

    def optimizer_step(self, optimizer, opt_idx, batch_idx, train_step_and_backward_closure):
        model_ref = self.trainer.get_model()

        is_lbfgs = isinstance(optimizer, torch.optim.LBFGS)
        using_native_amp = self.trainer.amp_backend == AMPType.NATIVE

        # native amp + lbfgs is a no go right now
        if using_native_amp and is_lbfgs:
            raise MisconfigurationException(
                'native PyTorch amp and lbfgs are not compatible.'
                ' To request, please file a Github issue in PyTorch and tag @mcarilli')

        # wraps into LightingOptimizer only for running step
        optimizer = LightningOptimizer._to_lightning_optimizer(optimizer, self.trainer, opt_idx)
```

```
    @classmethod
    def _to_lightning_optimizer(cls, optimizer, trainer, opt_idx):
        # apex overrides .step function and need to be wrapped on each step
        if trainer.amp_backend == AMPType.APEX:
            optimizer = cls(optimizer)
            optimizer._on_trainer_init(trainer)
        else:
            optimizer = trainer.lightning_optimizers[opt_idx]
        return optimizer
```

```
 if self._lightning_optimizers is None:
            self.convert_to_lightning_optimizers()
```

So, it is properly wrapped only when needed.

Reply:
Why do we set it to None in `__getstate__`? To avoid pickling it?